### Hello,
** In this notebook I'm working with the HackerNews dataset. Only 20,000 data rows are included.
I will analyze ASK and SHOW posts: **
* Rating difference
* Comments by time

In [1]:
from csv import reader
import numpy as np

In [2]:
open_file = open("hacker_news.csv")
read_file = reader(open_file)
hn = list(read_file)

In [3]:
print(hn[:5]) #has headers as first row

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [4]:
headers = hn[0] 
del hn[0] #deleting headers from hn list of lists
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


** Creating additional lists that only contain data for posts with ASK and SHOW **

In [5]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

print("Number of posts:")
print("Ask posts: ", len(ask_posts))
print("Show posts: ", len(show_posts))
print("Other posts: ",len(other_posts))

Number of posts:
Ask posts:  1744
Show posts:  1162
Other posts:  17194


** Observing comments and Average amount of comments for ASK and SHOW posts **

In [6]:
total_ask_comments = 0 #Total commments for ask_posts
for post in ask_posts:
    num_comments = int(post[4])
    total_ask_comments += num_comments
avg_ask_comments = total_ask_comments/len(ask_posts) #Avg of comments in the ask_posts variable 
print("Average amount of comments for Ask posts: ",avg_ask_comments) 

Average amount of comments for Ask posts:  14.038417431192661


In [7]:
total_show_comments = 0 #Total comments for show_posts
for post in show_posts:
    num_comments = int(post[4])
    total_show_comments += num_comments
    avg_show_comments = total_show_comments / len(show_posts)
print("Average amount of comments for Show posts: ",avg_show_comments) 

Average amount of comments for Show posts:  10.31669535283993


** Result : On average, ASK posts receive a higher amount of comments **

** Only ASK posts will be analyzed: **
* Amount of ASK posts per hour, Amount of comments received per hour
* Average comments per hour 

** Creating list with

In [8]:
import datetime as dt

result_list = [] #contains list [created at, amount of comments]
for post in ask_posts:
    created_at = post[6]
    comments = int(post[4])
    result_list.append([created_at, comments])    

** Creating Dictionaries with posts per hour and comments per hour **

In [9]:
counts_by_hour = {}
comments_by_hour = {}
for row in result_list:
    hour = row[0]
    comment = row[1]
    hour_dt = dt.datetime.strptime(hour, "%m/%d/%Y %H:%M")
    hour_str = hour_dt.strftime("%H")
    if hour_str not in counts_by_hour:
        counts_by_hour[hour_str] = 1
        comments_by_hour[hour_str] = comment
    else:
        counts_by_hour[hour_str] += 1
        comments_by_hour[hour_str] += comment

** Creating list of lists with hour, average comments **

In [10]:
avg_by_hour = []
for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])
avg_by_hour

[['10', 13.440677966101696],
 ['20', 21.525],
 ['14', 13.233644859813085],
 ['15', 38.5948275862069],
 ['19', 10.8],
 ['07', 7.852941176470588],
 ['03', 7.796296296296297],
 ['06', 9.022727272727273],
 ['04', 7.170212765957447],
 ['16', 16.796296296296298],
 ['12', 9.41095890410959],
 ['13', 14.741176470588234],
 ['00', 8.127272727272727],
 ['05', 10.08695652173913],
 ['02', 23.810344827586206],
 ['17', 11.46],
 ['22', 6.746478873239437],
 ['11', 11.051724137931034],
 ['23', 7.985294117647059],
 ['01', 11.383333333333333],
 ['21', 16.009174311926607],
 ['18', 13.20183486238532],
 ['09', 5.5777777777777775],
 ['08', 10.25]]

In [24]:
swap_avg_by_hour = []
for hour in avg_by_hour:
    swap_avg_by_hour.append([hour[1], hour[0]])
swap_avg_by_hour

[[13.440677966101696, '10'],
 [21.525, '20'],
 [13.233644859813085, '14'],
 [38.5948275862069, '15'],
 [10.8, '19'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [9.022727272727273, '06'],
 [7.170212765957447, '04'],
 [16.796296296296298, '16'],
 [9.41095890410959, '12'],
 [14.741176470588234, '13'],
 [8.127272727272727, '00'],
 [10.08695652173913, '05'],
 [23.810344827586206, '02'],
 [11.46, '17'],
 [6.746478873239437, '22'],
 [11.051724137931034, '11'],
 [7.985294117647059, '23'],
 [11.383333333333333, '01'],
 [16.009174311926607, '21'],
 [13.20183486238532, '18'],
 [5.5777777777777775, '09'],
 [10.25, '08']]

In [25]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
sorted_swap[:5]

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21']]

In [29]:
print("Top 5 Hours for Ask Posts Comments")
for avg, hour in sorted_swap[:5]:
    print("{}: {:.2f} average comments per post".format(dt.datetime.strptime(hour, "%H").strftime("%H:%M"), avg))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


** According to this list, the time when ask posts receive the most comments is at 3 pm. On average people comment about 38.59 per post **